File: apps/open-swe-v2/post_model_hook.py
"""
Post model hook for the coding agent with intelligent caching for command approvals.
Similar to Claude Code's approval system - only prompts for new commands or new directories.
"""

from typing import Dict, Any
from langchain_core.messages import AIMessage
from langgraph.types import interrupt
from .constants import WRITE_COMMANDS
from .state import CodingAgentState


def create_coding_agent_post_model_hook():
    """Create a post model hook with intelligent approval caching."""
    
    def post_model_hook(state: Dict[str, Any]) -> Dict[str, Any]:
        """
        Post model hook that checks for write tool calls and uses caching to avoid
        redundant approval prompts for the same command/directory combinations.
        """
        # Get the last message from the state
        messages = state.get("messages", [])
        if not messages:
            return state
        
        last_message = messages[-1]
        
        if not isinstance(last_message, AIMessage) or not last_message.tool_calls:
            return state
        
        # Create a CodingAgentState instance to access approval methods
        coding_state = CodingAgentState()
        coding_state.approved_operations = state.get("approved_operations", {"cached_approvals": set()})
        
        approved_tool_calls = []
        
        # Define write tools that need approval
        write_tools = WRITE_COMMANDS
        
        for tool_call in last_message.tool_calls:
            tool_name = tool_call.get("name", "")
            tool_args = tool_call.get("args", {})
            
            if tool_name in write_tools:
                # Check if this command/directory combination has been approved before
                if coding_state.is_operation_approved(tool_name, tool_args):
                    approved_tool_calls.append(tool_call)
                else:
                    approval_key = coding_state.get_approval_key(tool_name, tool_args)
                    
                    is_approved = interrupt({
                        "command": tool_name,
                        "args": tool_args,
                        "approval_key": approval_key
                    })
                    
                    if is_approved:
                        coding_state.add_approved_operation(tool_name, tool_args)
                        approved_tool_calls.append(tool_call)

                        state["approved_operations"] = coding_state.approved_operations
                    else:
                        continue
            else:
                approved_tool_calls.append(tool_call)
        
        # Update the message if any tool calls were filtered out
        if len(approved_tool_calls) != len(last_message.tool_calls):
            new_message = AIMessage(
                content=last_message.content,
                tool_calls=approved_tool_calls,
                additional_kwargs=last_message.additional_kwargs
            )
            
            new_messages = messages[:-1] + [new_message]
            state["messages"] = new_messages
        
        return state
    
    return post_model_hook

File: apps/open-swe-v2/src/post-model-hook.ts
import { AIMessage, AIMessageChunk } from "@langchain/core/messages";
import { interrupt } from "@langchain/langgraph";
import { WRITE_COMMANDS } from "./constants.js";
import { CodingAgentStateHelpers, type CodingAgentStateType } from "./state.js";

interface StateType extends CodingAgentStateType {
  todos: any[];
  files: Record<string, string>;
  messages: any[];
}

interface ToolCall {
  name?: string;
  args?: Record<string, any>;
  [key: string]: any;
}

export function createCodingAgentPostModelHook() {
  /**
   * Post model hook that checks for write tool calls and uses caching to avoid
   * redundant approval prompts for the same command/directory combinations.
   */
  async function postModelHook(
    state: StateType,
    _model?: any,
  ): Promise<StateType> {
    // Get the last message from the state
    const messages = state.messages || [];
    if (messages.length === 0) {
      return state;
    }

    const lastMessage = messages[messages.length - 1];

    if (
      !(
        lastMessage instanceof AIMessage ||
        lastMessage instanceof AIMessageChunk
      ) ||
      !lastMessage.tool_calls
    ) {
      return state;
    }

    if (!state.approved_operations) {
      state.approved_operations = { cached_approvals: new Set<string>() };
    }

    const approvedToolCalls: ToolCall[] = [];

    for (const toolCall of lastMessage.tool_calls) {
      const toolName = toolCall.name || "";
      const toolArgs = toolCall.args || {};

      // Skip tool calls without a name
      if (!toolCall.name) {
        continue;
      }

      if (WRITE_COMMANDS.has(toolName)) {
        // Check if this command/directory combination has been approved before
        if (
          CodingAgentStateHelpers.isOperationApproved(state, toolName, toolArgs)
        ) {
          approvedToolCalls.push(toolCall);
        } else {
          const approvalKey = CodingAgentStateHelpers.getApprovalKey(
            toolName,
            toolArgs,
          );

          const isApproved = interrupt({
            command: toolName,
            args: toolArgs,
            approval_key: approvalKey,
          });

          if (isApproved) {
            CodingAgentStateHelpers.addApprovedOperation(
              state,
              toolName,
              toolArgs,
            );
            approvedToolCalls.push(toolCall);
          } else {
            continue;
          }
        }
      } else {
        approvedToolCalls.push(toolCall);
      }
    }

    // Update the message if any tool calls were filtered out
    if (approvedToolCalls.length !== lastMessage.tool_calls.length) {
      const originalToolCalls = lastMessage.tool_calls.filter((toolCall) =>
        approvedToolCalls.some((approved) => approved.name === toolCall.name),
      );

      const MessageClass =
        lastMessage instanceof AIMessageChunk ? AIMessageChunk : AIMessage;
      const newMessage = new MessageClass({
        content: lastMessage.content,
        tool_calls: originalToolCalls,
        additional_kwargs: lastMessage.additional_kwargs,
      });

      const newMessages = [...messages.slice(0, -1), newMessage];
      state.messages = newMessages;
    }

    return state;
  }

  return postModelHook;
}
